ui/ops/src/dynamic/widgets/usage/usage.js

```diff
@@ -0,0 +1,192 @@
+          const groupData = records[dataset];
+          for (const record of groupData) {
+            if (!countsByEdge[record.x.getTime()])
+              countsByEdge[record.x.getTime()] = {};
+
+            countsByEdge[record.x.getTime()][dataset] = record.y;
+          }
+        }
+      }
+    }
+  }, {immediate: true});
+
+  return computed(() => {
+    return Object.entries(countsByEdge)
+        .map(([timestamp, groupData]) => ({
+          x: new Date(Number(timestamp)),
+          groups: groupData
+        }))
+        .sort((a, b) => a.x.getTime() - b.x.getTime());
+  });
+}
+
+/**
+ * Reads and calculates the total usage count for each span between the edges, returning it as a data series.
+ * The y property at index i in the response will be the total usage count for the period between edges[i] and edges[i+1].
+ * The last property at index i will be the most recent recorded usage count for the period between edges[i] and edges[i+1].
+ *
+ * @param {string} name
+ * @param {Date[]} edges
+ * @param {number | null} [countBefore] - the usage count before the first edge, if not null
+ * @return {Promise<{string: {x: Date, y: number, last: number}[]}>} - of size edges.length - 1
+ */
+async function readUsageCountSeries(name, edges, countBefore) {
+  const findEdges = (edges, at) => {
+    let i = binarySearch(edges, at, (a, b) => a.getTime() - b.getTime());
+    if (i < 0) {
+      // binarySearch will return the index _after_ the edge before the span as this is where the value would be inserted
+      i = ~i - 1;
+    }
+    const res = {beforeIdx: i, before: edges[i], after: null, afterIdx: null};
+    if (i < edges.length - 1) {
+      res.after = edges[i + 1];
+      res.afterIdx = i + 1;
+    }
+    return res;
+  }
+
+  /** @type {({x: Date, y: number, last: number}[])} */
+  const dates = Array(edges.length - 1).fill(null);
+  /** @type {({string: {x: Date, y: number, last: number}[]})} */
+  const dst = {};
+
+  const req = {
+    name,
+    period: {
+      startTime: edges[0],
+      endTime: edges[edges.length - 1],
+    },
+    pageSize: 1000,
+    orderBy: 'record_time DESC'
+  };
+
+  do {
+    const resp = await listAllocationsHistory(req, {});
+
+    if (resp.allocationRecordsList.length === 0) break;
+
+    let {before, after, beforeIdx} = findEdges(edges, timestampToDate(resp.allocationRecordsList[0].recordTime));
+
+    for (const record of resp.allocationRecordsList) {
+      let d = dst[record.allocation.groupId];
+      const recordTime = timestampToDate(record.recordTime);
+      if (recordTime < before || recordTime >= after) {
+        ({before, after, beforeIdx} = findEdges(edges, recordTime));
+        if (!after) break;
+      }
+      if (record.allocation.assignment > Allocation.Assignment.UNASSIGNED) {
+        if (!d)
+          d = [...dates]
+
+        if (d[beforeIdx])
+          d[beforeIdx].y++;
+        else
+          d[beforeIdx] = {x: before, y: 1, last: 0};
```

Comment:
Either add braces or put the statements on the same line. New lines after braceless if/else can get hard to read